src/cliffordsynthesis/encoding/TwoQubitEncoder.cpp

```diff
@@ -0,0 +1,251 @@
+        }
+      } else {
+        collectTwoQubitGateVariables(t / 2, q, true, twoQubitGateVariables);
+        collectTwoQubitGateVariables(t / 2, q, false, twoQubitGateVariables);
+        assertExactlyOne(twoQubitGateVariables);
+        IF_PLOG(plog::verbose) {
+          std::cout << twoQubitGateVariables.size() << std::endl;
+
+          TRACE() << "Two Qubit Gate variables at time " << t << " and qubit "
+                  << q;
+          for (const auto& var : twoQubitGateVariables) {
+            TRACE() << var.getName();
+          }
+        }
+      }
+    }
+  }
+}
+
+void encoding::TwoQubitEncoder::assertGateConstraints() {
+  DEBUG() << "Asserting gate constraints";
+  xorHelpers = logicbase::LogicMatrix{T + 1};
+  for (std::size_t t = 0U; t <= T; ++t) {
+    TRACE() << "Asserting gate constraints at time " << t;
+    splitXorR(tvars->r[t], t);
+    if (t == T) {
+      assertPauliGateConstraints(t);
+    } else if (t % 2 == 0) {
+      assertSingleQubitGateConstraints(t);
+    } else {
+      assertTwoQubitGateConstraints(t);
+    }
+    TRACE() << "Asserting r changes at time " << t;
+    lb->assertFormula(tvars->r[t + 1] == xorHelpers[t].back());
+  }
+}
+
+void encoding::TwoQubitEncoder::assertSingleQubitGateConstraints(
+    const std::size_t pos) {
+  for (std::size_t q = 0U; q < N; ++q) {
+    assertZConstraints(pos, q);
+    assertXConstraints(pos, q);
+    assertRConstraints(pos, q);
+  }
+}
+
+void TwoQubitEncoder::assertRConstraints(const std::size_t pos,
+                                         const std::size_t qubit) {
+  for (const auto gate : SINGLE_QUBIT_GATES) {
+    const auto& change =
+        LogicTerm::ite(vars.gS[pos / 2][gateToIndex(gate)][qubit],
+                       tvars->singleQubitRChange(pos, qubit, gate),
+                       LogicTerm(0, static_cast<std::int16_t>(S)));
+    splitXorR(change, pos);
+  }
+}
+
+void TwoQubitEncoder::assertGatesImplyTransform(
+    const std::size_t pos, const std::size_t qubit,
+    const std::vector<TransformationFamily>& transformations) {
+  const auto& singleQubitGates = vars.gS[pos / 2];
+  for (const auto& [transformation, gates] : transformations) {
+    auto gateOr = LogicTerm(false);
+    for (const auto& gate : gates) {
+      gateOr = gateOr || singleQubitGates[gateToIndex(gate)][qubit];
+    }
+    lb->assertFormula(LogicTerm::implies(gateOr, transformation));
+  }
+}
+
+void encoding::TwoQubitEncoder::assertPauliGateConstraints(
+    const std::size_t pos) {
+  for (std::size_t q = 0U; q < N; ++q) {
+    for (const auto gate : PAULI_QUBIT_GATES) {
+      const auto& change =
+          LogicTerm::ite(vars.gP[gateToIndex(gate)][q],
+                         tvars->singleQubitRChange(pos, q, gate),
+                         LogicTerm(0, static_cast<std::int16_t>(S)));
+
+      splitXorR(change, pos);
+    }
+    lb->assertFormula(tvars->x[pos][q] == tvars->x[pos + 1][q]);
+    lb->assertFormula(tvars->z[pos][q] == tvars->z[pos + 1][q]);
```

Comment:
Not necessary, but since only the phase is adjusted here, one could save on Tableau Variables.